libs/MeshKernel/src/Mesh2D.cpp

```diff
@@ -1934,3 +1934,645 @@ std::vector<int> Mesh2D::NodeMaskFromPolygon(const Polygons& polygon, bool insid
+                }
+
+                if (endNode != constants::missing::uintValue)
+                {
+                    irregularEndNodes[i] = endNode;
+                }
+            }
+        }
+    }
+}
+
+void Mesh2D::GatherNodesToMerge(const UInt startNode, const UInt endNode, const Edge& boundaryEdge, std::vector<std::pair<UInt, UInt>>& nodesToMerge, std::vector<int>& mergeIndicator) const
+{
+    if (startNode != constants::missing::uintValue && mergeIndicator[boundaryEdge.first] == 1)
+    {
+        nodesToMerge.push_back({boundaryEdge.first, startNode});
+        mergeIndicator[boundaryEdge.first] = -1;
+    }
+
+    if (endNode != constants::missing::uintValue && mergeIndicator[boundaryEdge.second] == 1)
+    {
+        nodesToMerge.push_back({boundaryEdge.second, endNode});
+        mergeIndicator[boundaryEdge.second] = -1;
+    }
+}
+
+void Mesh2D::GatherHangingNodes(const UInt primaryStartNode, const UInt primaryEndNode, const Edge& irregularEdge, std::vector<UInt>& hangingNodesOnEdge, UInt& numberOfHangingNodes, std::vector<int>& mergeIndicator) const
+{
+    UInt secondaryStartNode = irregularEdge.first;
+    UInt secondaryEndNode = irregularEdge.second;
+
+    if (mergeIndicator[secondaryStartNode] == 1 && secondaryStartNode != primaryStartNode && secondaryStartNode != primaryEndNode)
+    {
+        hangingNodesOnEdge[numberOfHangingNodes] = secondaryStartNode;
+        ++numberOfHangingNodes;
+        mergeIndicator[secondaryStartNode] = 0;
+    }
+
+    if (mergeIndicator[secondaryEndNode] == 1 && secondaryEndNode != primaryStartNode && secondaryEndNode != primaryEndNode)
+    {
+        hangingNodesOnEdge[numberOfHangingNodes] = secondaryEndNode;
+        ++numberOfHangingNodes;
+        mergeIndicator[secondaryEndNode] = 0;
+    }
+}
+
+void Mesh2D::ConnectCurvilinearQuadsDDType()
+{
+    // Only here to shorten several of the initialisations below
+    static constexpr UInt missingValue = constants::missing::uintValue;
+
+    // Elements with no neighbour
+    std::vector<UInt> elementsOnDomainBoundary;
+    elementsOnDomainBoundary.reserve(GetNumEdges());
+
+    // Edges with no neighbour
+    std::vector<UInt> edgesOnDomainBoundary;
+    edgesOnDomainBoundary.reserve(GetNumEdges());
+
+    GetElementsOnDomainBoundary(elementsOnDomainBoundary, edgesOnDomainBoundary);
+
+    std::vector<UInt> irregularEdgeCount(GetNumEdges(), missingValue);
+    std::vector<HangingNodeIndexArray> irregularEdge(GetNumEdges(), {missingValue, missingValue, missingValue, missingValue, missingValue});
+    std::vector<UInt> irregularStartNodes(GetNumEdges(), missingValue);
+    std::vector<UInt> irregularEndNodes(GetNumEdges(), missingValue);
+
+    GatherHangingNodeIds(edgesOnDomainBoundary, irregularEdge, irregularEdgeCount, irregularStartNodes, irregularEndNodes);
+
+    std::vector<bool> adjacentEdgeIndicator(GetNumEdges(), true);
+    std::vector<std::pair<UInt, UInt>> nodesToMerge;
+    nodesToMerge.reserve(4 * GetNumEdges());
```

Comment:
maybe use a named constant for 4. you can also put this variable in constants.hpp, namespace numerics

Reply:
Used an already existing constant in the class.
I do not think this should be a globally visible constant, as it is only of use for this algorithm.